torch_xla/experimental/scan.py

```diff
@@ -82,14 +88,335 @@ def scan(fn, init, xs):
+    for i in range(len(carry)):
+      fn_carry_history[i] = dynamic_update_slice(fn_carry_history[i], carry[i],
+                                                 idx)
+    result = call_fn_computation(carry, x, additional_inputs)
+    for i in range(flat_carry_len):
+      carry[i] = result.get_tuple_element(i)
+    for i in range(flat_y_len):
+      y = result.get_tuple_element(i + flat_carry_len)
+      ys[i] = dynamic_update_slice(ys[i], y, idx)
+    return (num_iters - one,) + pack(carry, xs, fn_carry_history, ys)
+
+  res = xb.Op.mkwhile(builder.params(), cond_fn, body_fn)
+  computation = res.build('scan')
+
+  outputs = torch_xla._XLAC._xla_user_computation('xla::scan',
+                                                  builder.param_tensors(),
+                                                  computation)
+  # skip the last num_additional_inputs
+  outputs = outputs[:len(outputs) - num_additional_inputs]
+  # `1:` to skip `num_iters`
+  carry, xs, fn_carry_history, ys = unpack(outputs[1:])
+
+  # Unflatten tensors back to PyTrees
+  return tree_unflatten(carry, carry_spec), tree_unflatten(
+      fn_carry_history, carry_spec), tree_unflatten(ys, fn_output_y_spec)
+
+
+@pytreeify
+class Scan(torch.autograd.Function):
+
+  @staticmethod
+  def forward(ctx, fn, init, xs):
+    # Forward pass, save inputs for backward
+    ctx._fn = fn
+    with torch._C._AutoDispatchBelowAutograd():
+      carry, carry_history, ys = _scan_impl(fn, init, xs)
+    flat_carry_history, carry_spec = tree_flatten(carry_history)
+    flat_xs, xs_spec = tree_flatten(xs)
+    ctx.save_for_backward(*flat_carry_history, *flat_xs)
+    ctx._flat_carry_len = len(flat_carry_history)
+    ctx._carry_spec = carry_spec
+    ctx._xs_spec = xs_spec
+    return carry, carry_history, ys
+
+  @staticmethod
+  def backward(ctx, grad_carry, grad_carry_history, grad_ys):
+    fn = ctx._fn
+    flat_carry_len = ctx._flat_carry_len
+    carry_spec = ctx._carry_spec
+    xs_spec = ctx._xs_spec
+    tensors_list = ctx.saved_tensors
+    carry_history = tree_unflatten(tensors_list[:flat_carry_len], carry_spec)
+    xs = tree_unflatten(tensors_list[flat_carry_len:], xs_spec)
+
+    def detach_tensor(inp: torch.Tensor) -> torch.Tensor:
+      x = inp.clone().detach()
+      x.requires_grad = inp.requires_grad
+      return x
+
+    def step_fn(grad_carry, pytree: Tuple[torch.Tensor, torch.Tensor,
+                                          torch.Tensor]):
+      grad_y, carry, x = pytree
+      # Compute the backward of a single scan iteration
+      detached_inputs = tree_map(detach_tensor, (carry, x))
+      with torch.enable_grad():
+        # `outputs` is a tuple of two PyTrees, the carry and y.
+        # In order to use `autograd.backward`, we need to flatten all
+        # PyTrees. The flattened `grad_carry` should match up with `carry`
+        # and similar for `y`.
+        outputs = fn(*detached_inputs)
+        output_carry, output_y = outputs
```

Comment:
wait... you are retracing the fwd here?

Reply:
Correct. I'm not sure if the CSE pass can combine this with the same fn in the fwd pass. If it can't, then the bwd of scan will be slower.

Reply:
ehh, what you do here is pretty much force the gradident accumulation(through most likely cancel by the CSE since there is no optimization barrier), this sounds like a bad idea

Reply:
I believe the barriers won't be needed. This is the recommendation from JAX, which says you don't want to wrap inputs into a barrier if the checkpointed function is to be used in a `scan`.

[1]: https://jax.readthedocs.io/en/latest/_autosummary/jax.checkpoint.html

Reply:
See `prevent_cse` documentation in https://jax.readthedocs.io/en/latest/_autosummary/jax.checkpoint.html

Reply:
This is now avoided since we trace `fn` symbolically into separate fwd and bwd graphs, and weave the bwd into the backward of scan.